contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity =0.8.16;
+
+import {IWhitelist} from "../../libraries/common/IWhitelist.sol";
+import {IL1MessageQueue} from "./IL1MessageQueue.sol";
+import {IL1MessageQueueWithGasPriceOracle} from "./IL1MessageQueueWithGasPriceOracle.sol";
+import {IL2GasPriceOracle} from "./IL2GasPriceOracle.sol";
+
+import {L1MessageQueue} from "./L1MessageQueue.sol";
+
+contract L1MessageQueueWithGasPriceOracle is L1MessageQueue, IL1MessageQueueWithGasPriceOracle {
+    /*************
+     * Constants *
+     *************/
+
+    /// @notice The intrinsic gas for transaction.
+    uint256 private constant INTRINSIC_GAS_TX = 21000;
+
+    /// @notice The intrinsic gas for each nonzero byte.
+    uint256 private constant INTRINSIC_GAS_NONZERO_BYTE = 16;
+
+    /*************
+     * Variables *
+     *************/
+
+    /// @notice The latest known l2 base fee.
+    uint256 public l2BaseFee;
+
+    /// @notice The address of whitelist contract.
+    address public whitelist;
+
+    /***************
+     * Constructor *
+     ***************/
+
+    /// @notice Constructor for `L1MessageQueueWithGasPriceOracle` implementation contract.
+    ///
+    /// @param _messenger The address of `L1ScrollMessenger` contract.
+    /// @param _scrollChain The address of `ScrollChain` contract.
+    /// @param _enforcedTxGateway The address of `EnforcedTxGateway` contract.
+    constructor(
+        address _messenger,
+        address _scrollChain,
+        address _enforcedTxGateway
+    ) L1MessageQueue(_messenger, _scrollChain, _enforcedTxGateway) {}
+
+    /// @notice Initialize the storage of L1MessageQueueWithGasPriceOracle.
+    function initializeV2() external reinitializer(2) {
```

Comment:
Should we ensure that the functions below can only be called once `initializeV2` has been called?

Reply:
not necessary, will introduce extra gas costs.